src/gui_common/TweakedColourPicker.cs

```diff
@@ -0,0 +1,340 @@
+                AddPreset(color);
+            PresetsStorage.Remove(presetsStored);
+        }
+
+        UpdateTooltips();
+    }
+
+    public override void _ExitTree()
+    {
+        // Store presets.
+        PresetsStorage.Add(new KeyValuePair<string, Color[]>(GetPath(), GetPresets()));
+
+        base._ExitTree();
+    }
+
+    public override void _Notification(int what)
+    {
+        if (what == NotificationTranslationChanged)
+            UpdateTooltips();
+
+        base._Notification(what);
+    }
+
+    public new void AddPreset(Color color)
+    {
+        if (presets.Any(knownPreset => knownPreset.Color == color))
+            return;
+
+        // Add preset locally
+        var preset = new TweakedColorPickerPreset(this, color);
+        presets.Add(preset);
+        presetsContainer.AddChild(preset);
+
+        // Add preset to base class
+        base.AddPreset(color);
+    }
+
+    public new void ErasePreset(Color color)
+    {
+        OnPresetDeleted(presets.First(p => p.Color == color));
+    }
+
+    private void UpdateTooltips()
+    {
+        pickerButton.HintTooltip = TranslationServer.Translate("COLOR_PICKER_PICK_COLOR");
+        addPresetButton.HintTooltip = TranslationServer.Translate("COLOR_PICKER_ADD_PRESET");
+    }
+
+    private void UpdateButtonsState()
+    {
+        if (hsvCheckButton == null)
+            return;
+
+        hsvCheckButton.Pressed = HsvMode;
+        rawCheckButton.Pressed = RawMode;
+
+        hsvCheckButton.Disabled = !hsvButtonEnabled || RawMode;
+        rawCheckButton.Disabled = !rawButtonEnabled || HsvMode;
+    }
+
+    private void OnAddPresetButtonPressed()
+    {
+        AddPreset(Color);
+    }
+
+    private void OnPresetSelected(Color color)
+    {
+        Color = color;
+        EmitSignal("color_changed", Color);
+    }
+
+    private void OnPresetDeleted(TweakedColorPickerPreset preset)
+    {
+        presets.Remove(preset);
+        presetsContainer.RemoveChild(preset);
+        base.ErasePreset(preset.Color);
+        preset.QueueFree();
+    }
+
+    private void OnHSVButtonToggled(bool isOn)
+    {
+        HsvMode = isOn;
+    }
+
+    private void OnRawButtonToggled(bool isOn)
+    {
+        RawMode = isOn;
+    }
+
+    private void OnColorChanged(Color color)
+    {
+        hexColorEdit.Text = color.a8 == 255 ? color.ToHtml(false) : color.ToHtml(true);
+    }
+
+    /// <summary>
+    ///   Called when HexColorEdit's text has changed (because of user input / color select). In this function,
+    ///   the input has been confirmed, so if we find something wrong (such as illegal input), we should delete it
+    ///   and set the text to a legal modified one. And because caret position is set to 0 after a '.Text = ""', we
+    ///   need to manually adjust the caret position to avoid strange behavior.
+    /// </summary>
+    /// <param name="color">HexColorEdit's text content</param>
+    private void OnHexColorChanged(string color)
```

Comment:
It'd actually be more intuitive if the color is only applied after the text edit is entered in my opinion (`text_entered`), the default color picker behavior follows the same behavior. This will also allows for simple html validity check approach and no need to manually adjusting the caret, if the color code is invalid we can just force a default value like `00000000` to be applied to the color picker and text edit.

Note that with this apply-on-enter approach, if you focus out of the text edit (selecting other element while editing the textbox), the color may not be applied and kind of just stuck. In order to solve this you need to manually apply the color on focus exit using signal, see [color_picker.cpp#L660](https://github.com/godotengine/godot/blob/3.3/scene/gui/color_picker.cpp#L660).

Reply:
But you can't let any non-hex characters exist there. It's against the edit's usage (as a hex input). Anyway I think that it is necessary to ensure there isn't any illegal characters.

Reply:
The text edit still doesn't allow non-hex characters but _only_ after the text edit is entered using <kbd>Enter</kbd>. If there is any illegal characters or the HTML hexadecimal color string is invalid on enter using the validity check, the text in hex input field and applied color is then replaced with the mentioned default fallback value.